src/garage/torch/modules/cnn_module.py

```diff
@@ -0,0 +1,276 @@
+
+            hidden_layers = nn.Sequential()
+
+            if isinstance(paddings, (list, tuple)):
+                padding = paddings[index]
+
+            # conv 2D layer
+            conv_layer = _conv(in_channels=prev_channel,
+                               out_channels=channel,
+                               kernel_size=kernel_size,
+                               stride=stride,
+                               padding=padding)
+            hidden_w_init(conv_layer.weight)
+            hidden_b_init(conv_layer.bias)
+            hidden_layers.add_module('conv_{}'.format(index), conv_layer)
+
+            # layer normalization
+            if layer_normalization:
+                hidden_layers.add_module('layer_normalization',
+                                         nn.LayerNorm(channel))
+
+            # non-linear function
+            if hidden_nonlinearity:
+                hidden_layers.add_module('non_linearity',
+                                         _NonLinearity(hidden_nonlinearity))
+            # max-pooling
+            if max_pool:
+                hidden_layers.add_module(
+                    'max_pooling',
+                    nn.MaxPool2d(kernel_size=pool_shapes, stride=pool_strides))
+
+            self._layers.append(hidden_layers)
+            prev_channel = channel
+
+    @classmethod
+    def _check_parameter_for_output_layer(cls, var, n_layers):
+        """Check input parameters for conv layer are valid.
+
+        Args:
+            var (any): variable to be checked
+            n_layers (int): number of layers
+
+        Returns:
+            list: list of variables (length of n_layers)
+
+        Raises:
+            ValueError: if the variable is a list but length of the variable
+                is not equal to n_layers
+
+        """
+        if isinstance(var, (list, tuple)):
+            if len(var) == 1:
+                return list(var) * n_layers
+            if len(var) == n_layers:
+                return var
+            msg = ('{} should be either an integer or a collection of length '
+                   'n_layers ({}), but got {} instead.')
+            raise ValueError(msg.format(str(var), n_layers, var))
+        return [copy.deepcopy(var) for _ in range(n_layers)]
+
+    @classmethod
+    def _check_valid_padding(cls, paddings, filter_sizes):
+        """Check if padding shape of convolutional layers are valid for `same`\
+        padding (zero padding) implementation.
+
+        Args:
+            paddings (tuple[int]): paddings
+            filter_sizes (tuple[int]): filter sizes
+
+        Raises:
+            ValueError: if the padding size is not equal to the expected shape
+
+        """
+        for index, (padding,
+                    filter_size) in enumerate(zip(paddings, filter_sizes)):
+            if padding != 0:  # only check when padding value is not 0
+                expected_padding = int((filter_size - 1) / 2)  # P = (F - 1) /2
```

Comment:
Why does padding and kernel size need to follow this restriction anyway?

Reply:
My initial idea is to set zero padding to be P=(F−1)/2 so that when the padding mode is `zero`, it will check whether the padding size fits this formula. 

But it is only useful in a marginal case when stride=1. Now I think I'll remove this function.